### DISCUSSION TITLE
How to block ligation or mark-to-mark attachment between two marks

### DISCUSSION THREAD

**Discussion by khaledhosny** [harfbuzz/harfbuzz#4796] [Q&A] [answered]
Lets say a font has a _shadda_+_kasra_ ligature or mark-to-mark attachment between the two glyphs, but one wants to block that at text input level, so that the _kasra_ is attached to the base glyph?

My fist intuition was using CGJ, but it block only mark-to-mark attachment and not ligatures:
```
# this font uses mark-to-mark attachment
$ hb-view "Fustat[wght].ttf" -u 0628,0651,034F,0650,0627,0628,0650,0627 
```
![fustat-cgj](https://github.com/harfbuzz/harfbuzz/assets/93914/f90d251b-a090-458f-8ae2-3c45c263c534)

```
# this font uses ligatures
$ hb-view "NotoSansArabic[wdth,wght].ttf" -u 0628,0651,034F,0650,0627,0628,0650,0627
```
![nsa-cgj](https://github.com/harfbuzz/harfbuzz/assets/93914/c9920a8a-9251-490d-82e3-5f7e709cfec5)

(Interestingly, in CoreText CGJ blocks both, but in Uniscribe while it blocks both, it also break Arabic joining) 

Next, I tried ZWNJ, which block both mark-to-mark attachment and ligatures, but obviously breaks Arabic joining:
```
$ hb-view "Fustat[wght].ttf" -u 0628,0651,200C,0650,0627,0628,0650,0627
```
![fustat-zwnj](https://github.com/harfbuzz/harfbuzz/assets/93914/da4826a9-b435-40e8-bf8f-217d9b420fff)

```
$ hb-view "NotoSansArabic[wdth,wght].ttf" -u 0628,0651,200C,0650,0627,0628,0650,0627
```
![nsa-zwnj](https://github.com/harfbuzz/harfbuzz/assets/93914/ae7808bf-fa87-4c78-b9e2-c7cfedcdcbeb)

Last I tried, ZWJ, and it breaks both and mark-to-base attachment for the _kasra_, and obviously forces Arabic joining:

```
$ hb-view "Fustat[wght].ttf" -u 0628,0651,200D,0650,20,0628,0650

```
![fustat-zwj](https://github.com/harfbuzz/harfbuzz/assets/93914/aed4d09a-2cf1-42b8-b2da-b1b71195fdfd)

```
$ hb-view "NotoSansArabic[wdth,wght].ttf" -u 0628,0651,200D,0650,20,0628,0650
```
![nsa-zwj](https://github.com/harfbuzz/harfbuzz/assets/93914/4a204a61-422e-4870-b6e8-a15497621b1e)

**Comment by behdad** [comment]
I'm not sure what needs to be done here. cc @tiroj @roozbehp

**Reply by khaledhosny** [reply]
Matching CoreText behavior here (CGJ block both mark-to-mark attachment and ligation) seems sensible to me, unless it breaks something else.

**Reply by khaledhosny** [reply, marked as answer]
It just occurred to me that I was approaching this the wrong way. I need to input _kasra_ + CGJ + _shadda_ in that order, so that CGJ prevents the canonical mark re-ordering which would autoatically block ligation or mark-to-mark attachment without the layout engine having to treat it in any other special way. So the above test strings should be.

This works partially in HarfBuzz, though. It prevents ligation just fine:
```
$ hb-view "NotoSansArabic[wdth,wght].ttf" -u 0628,0650,034F,0651
```
![noto-sans-arabic](https://github.com/user-attachments/assets/a7705f31-f4e5-4477-8d5f-80c7f47b8d71)

It also prevents mark-to-mark attachment, but then the shadda is not attached to the base glyph:
```
$ hb-view "Fustat[wght].ttf" -u 0628,0650,034F,0651
```
![fustat](https://github.com/user-attachments/assets/8df75f14-2361-4f47-8043-f3c327eb299e)

In CoreText the shadda attaches to the base just fine:
```
$ hb-view "Fustat[wght].ttf" -u 0628,0650,034F,0651 --shaper=coretext
```
![fustat-ct](https://github.com/user-attachments/assets/9523ca5e-daa1-45be-b892-f4e2d1b79287)

I think this can be considered a bug in HarfBuzz. The presence of CGJ should prevent the re-ordering (it does), but it shouldn’t prevent mark-to-base attachment.

**Comment by tiroj** [comment]
According to TUS, CGJ can be used to prevent mark reordering in normalisation. So _potentially_ someone could use it in that way while not wanting it to interfere with mark interaction in terms of positioning or ligating.

[CGJ has a canonical combining class of zero, which made it useful for preventing normalisation reordering. This behaviour was formally recognised to address issues in Hebrew canonical combining classes that produced incorrect ordering in some specific Biblical text editions.]

**Reply by khaledhosny** [reply]
> So potentially someone could use it in that way while not wanting it to interfere with mark interaction in terms of positioning or ligating.

CGJ already blocks ligation and mark-to-mark positioning in CoreText and DirectWrite (though DirectWrite behavior is completely broken). In HarfBuzz it blocks mark-to-mark positioning but allows ligation. So it is unlikely that anyone currently depends on such behavior.

Though, I don’t really care about CGJ itself, I only want a way to force _kasra_ to always attach to base glyph, which is not uncommon thing to want (for example https://github.com/notofonts/arabic/issues/242, and also for Quranic text where current Mushafs always show the _kasra_ below the base glyph).

**Comment by tiroj** [comment]
Presumably, if someone were working with a script where CGJ was only intended to prevent reordering but not to break mark interaction, that could be addressed explicitly at the GSUB level, either by including CGJ in ligature lookups or stripping it from the glyph string.

As an aside, I have thought for a long time that Unicode would have benefited from mark formatting control characters, with a distinction between formatting controls affecting base characters and those affecting marks.

**Comment by tiroj** [comment]
[In the Biblical Hebrew case, CGJ is used to prevent reordering of the _meteg_ U+05BD, which normally sits to the left of a vowel, but in some editions may also occur to the right of the vowel, and with some vowels conventionally forms a ligature in which it sits between the two parts of a (atomically encoded) _hataf_ vowel. Due to uncorrectable errors in the Hebrew canonical combining class assignments, normalisation prevents the right-side _meteg_ positioning, so CGJ can be used to prevent that reordering. But the resulting mark sequence still needs to interact for mark-to-mark positioning (and context for mark-to-base repositioning of the vowel). I will need to go back and look at the SBL Hebrew source to remember how I did this, but I suspect I stripped CGJ from the glyph string after it had ceased being useful in GSUB.]

**Comment by roozbehp** [comment]
As @tiroj mentioned, CGJ is used mostly to stop the normalization algorithm from reordering combining marks. I believe it should not block mark-to-mark positioning. In Quranic Arabic, for example, it's important that both the sequences <dhal, kasra, U+08D9> and <dhal, U+08D9, CGJ, kasra> work, result in different renderings, and (I believe) have mark-to-mark positioning working for them. (See https://www.unicode.org/reports/tr53/proposed.html#Mistaken for some details and explanation.)
